lib/mongo/auth/stringprep.rb

```diff
@@ -0,0 +1,68 @@
+require 'mongo/auth/stringprep/tables'
+
+module Mongo
+  module Auth
+    module StringPrep
+      extend self
+
+      def prepare(data, mappings, prohibited, options = {})
+        out = apply_maps(data, mappings)
+        normalize(out) if options[:normalize]
+        check_prohibited(out, prohibited)
+        check_bidi(out) if options[:bidi]
+        out
+      end
+
+      private
+
+      def apply_maps(data, mappings)
+        out = ''
+
+        data.each_char do |c|
+          out << mapping(c, mappings)
+        end
+
+        out
+      end
+
+      def check_bidi(out)
+        if out.any? { |c| table_contains?(Tables::C8, c) }
+          raise Error::StringPrep(Error::StringPrep::INVALID_BIDIRECTIONAL)
+        end
+
+        if out.any? { |c| table_contains?(Tables::D1, c) }
+          unless table_contains?(Tables::D1, out.first) && table_contains?(Tables::D1, out.last)
```

Comment:
Maybe I'm missing this, but what's the point of using #any? if you raise an error as soon as the condition is not met?

Reply:
The algorithm is that if any character in the string is in table D1, then it's only valid if the first and last characters are also in D1. Using `any?` seemed like the most straightforward way to implement this.

Reply:
Oh, wait, I see, you're saying I should use `each`. That's a mistaken holdover from before I implemented the error throwing; I'll fix it.